datafusion/functions/src/string/common.rs

```diff
@@ -174,3 +177,62 @@ where
         },
     }
 }
+
+pub(crate) fn case_conversion<'a, F>(
+    args: &'a [ColumnarValue],
+    op: F,
+    name: &str,
+) -> Result<ColumnarValue>
+where
+    F: Fn(&'a str) -> String,
+{
+    match &args[0] {
+        ColumnarValue::Array(array) => match array.data_type() {
+            DataType::Utf8 => {
+                Ok(ColumnarValue::Array(convert_array::<i32, _>(array, op)?))
+            }
+            DataType::LargeUtf8 => {
+                Ok(ColumnarValue::Array(convert_array::<i64, _>(array, op)?))
+            }
+            other => exec_err!("Unsupported data type {other:?} for function {name}"),
+        },
+        ColumnarValue::Scalar(scalar) => match scalar {
+            ScalarValue::Utf8(a) => {
+                let result = a.as_ref().map(|x| op(x));
+                Ok(ColumnarValue::Scalar(ScalarValue::Utf8(result)))
+            }
+            ScalarValue::LargeUtf8(a) => {
+                let result = a.as_ref().map(|x| op(x));
+                Ok(ColumnarValue::Scalar(ScalarValue::LargeUtf8(result)))
+            }
+            other => exec_err!("Unsupported data type {other:?} for function {name}"),
+        },
+    }
+}
+
+fn convert_array<'a, O, F>(array: &'a ArrayRef, op: F) -> Result<ArrayRef>
+where
+    O: OffsetSizeTrait,
+    F: Fn(&'a str) -> String,
+{
+    let string_array = as_generic_string_array::<O>(array)?;
+    let value_data = string_array.value_data();
+
+    // SAFETY: all items stored in value_data satisfy UTF8.
+    // ref: impl ByteArrayNativeType for str {...}
+    let str_values = unsafe { std::str::from_utf8_unchecked(value_data) };
+
+    // conversion
+    let converted_values = op(str_values);
+    let bytes = converted_values.into_bytes();
+
+    // build result
+    let values = Buffer::from_vec(bytes);
+    let offsets = string_array.offsets().clone();
+    let nulls = string_array.nulls().cloned();
+
+    // SAFETY: offsets and nulls are consistent with the input array.
```

Comment:
I think the safety of this also relies on `op` not decreasing the length of the characters, as in that case the offsets would be pointing into invalid characters. For example if `op()` always returned `""` I think this would result in an invalid array.

Can you please also add a sanity check that the the new string didn't get smaller?

Something like (untested) this above:
```rust
    let orig_len = str_values.len()
    let converted_values = op(str_values);
    // offsets may refer to anywhere in str_values so new string must be at least as long
    assert!(converted_values.len() >= orig_len);